### ISSUE TITLE
Attribute prefix and suffix selector should match nothing if the string is empty

### ISSUE THREAD

**Issue by upsuper** [servo/servo#15394]
See the following code:
```html
<!DOCTYPE html>
<style>
  p { color: green }
  p[attr$=""] { color: red }
  p[attr^=""] { color: red }
</style>
<p attr="foo">hello</p>
```

Both attribute selectors should match nothing, so the color should be green.

Spec: https://drafts.csswg.org/selectors-4/#attribute-substrings

This is probably an issue in servo/rust-selectors, and might be an easy issue.

cc @SimonSapin

**Comment by SimonSapin** [comment]
This needs to be fixed in implementations of the `match_attr_prefix` and `match_attr_suffix` trait methods. We have two of them:

* https://github.com/servo/rust-selectors/blob/master/src/tree.rs
* https://github.com/servo/servo/blob/master/components/style/gecko/wrapper.rs

Alternatively, we could change the selector parser to emit new `SimpleSelector::AttrPrefixEmpty` and `SimpleSelector::AttrSuffixEmpty` variants, and have `matches_simple_selector` immediately return `false` for these variants.

@upsuper, which do you prefer?

**Comment by upsuper** [comment]
I think it is probably better being handled inside selectors crate, since it is part of the selectors spec, not something depends on the host language, and thus I don't think the behavior should depend on how the trait methods are implemented.

**Comment by SimonSapin** [comment]
I’m working on this. Substring and "includes" attr selectors are also affected:

```html
<!DOCTYPE html>
<style>
  p { color: green }
  p[attr~=""] { color: red }
  p[attr^=""] { color: red }
  p[attr*=""] { color: red }
  p[attr$=""] { color: red }
</style>
<p attr="foo ">hello</p>
```

(The space in `attr="foo "` is important, since `"foo ".split(' ') == ["foo", ""]` and we get an empty string for `~=` to (incorrectly) match.)

**Comment by SimonSapin** [comment]
rust-selectors v0.16.0 is on crates.io and should fix this. What remains to do is update servo to it and add tests.

**Comment by emilio** [comment]
This is fixed now, filled  #15448 for the tests.
